Upgrade to ipfs@0.41 (#381)
ProtoSchool/protoschool.github.io: src/tutorials/0005-regular-files-api/04.md

```javascript
ipfs.cat(ipfsPath, [options])
```

An IPFS `path` can take several forms (you can read about them in the [Files API](https://github.com/ipfs/interface-js-ipfs-core/blob/master/SPEC/FILES.md#cat) documentation). In this lesson, we'll use the simplest one: the `cid` present in the object we receive for a file when the `ipfs.add` method is called. You may remember from the previous lesson that the `add` method returned both `path` and `cid` values for each file you added to IPFS. We'll learn more about the difference between these two values later.

The `cat` method first searches your own node for the file requested, and if it can't find it there, it will attempt to find it on the broader IPFS network. Because of the way cryptographic hashing works, the content you're searching for is guaranteed to match the file found, regardless of which peer is hosting it.

> The `cat` method returns an Async Iterable, that iterates over the file's chunks of data, i.e. Buffers.

Reviewer: ```suggestion
The `cat` method returns an Async Iterable that iterates over the file's chunks of data, i.e. Buffers.
```

A `Buffer` is just a raw collection of bytes and as such doesn't make any assumptions about encodings or the type of data it contains. However, if we know the file being retrieved is a plain text file such as a `.txt`, we can convert its buffered contents to a UTF-8 string (an interpretation of those raw bytes) by calling the JavaScript method `.toString()`.

So if you had the CID for a text file in an IPFS node, you could retrieve the file's contents as a readable string like so:

> 

Reviewer: Looks like we're missing an intro of it-concat here

```javascript
// the concat variable is globally available (just like ipfs)

const bufferedContents = await concat(ipfs.cat('QmWCscor6qWPdx53zEQmZvQvuWQYxx1ARRCXwYVE4s9wzJ')) // returns a Buffer
const stringContents = bufferedContents.toString() // returns a string
```

When you're ready to try this in the real world, you should note that the above example can result in heavy memory usage, depending on the contents of the file being read. If you find this to be the case, you might want to skip the usage of the `it-concat` package and process each chunk of data iteratively.
This is the main reason why ipfs returns `Async Iterables` so that we can decide if we want to be aware of memory performance issues or not.

> In ProtoSchool tutorials, we don't really need to worry about this since our code challenges use small files, so we can concatenate everything without worrying about performance.

Reviewer: ```suggestion
In ProtoSchool tutorials, our code challenges use small files, so we can concatenate everything without worrying about performance.
```

A `Buffer` is just a raw collection of bytes and as such doesn't make any assumptions about encodings or the type of data it contains. However, if we know the file being retrieved is a plain text file such as a `.txt`, we can convert its buffered contents to a UTF-8 string (an interpretation of those raw bytes) by calling the JavaScript method `.toString()`.

So if you had the CID for a text file in an IPFS node, you could retrieve the file's contents as a readable string like so:

```javascript
const concat = require('it-concat')

const bufferedContents = await concat(ipfs.cat('QmWCscor6qWPdx53zEQmZvQvuWQYxx1ARRCXwYVE4s9wzJ')) // returns a Buffer
const stringContents = bufferedContents.toString() // returns a string
```

Notice how the example above waits for the concatenation of the results of the `cat` method (using the [`it-concat`](https://www.npmjs.com/package/it-concat) package) before converting the value to a string. To accomplish that in a single line of code, you'd need to wrap the `await` statement in parentheses like so:

```javascript
const stringContents = (await concat(ipfs.cat(cid))).toString() // returns a string

> ```

Reviewer 2: I don't think these single line versions add any value to the turorial. Saving a line of code should not be a concern for a newcomer (or arguably any developer).

When you're ready to try this in the real world, you should note that the above example can result in heavy memory usage, depending on the contents of the file being read. If you find this to be the case, you might want to skip the usage of the `it-concat` package and process each chunk of data iteratively.
This is the main reason why ipfs returns `Async Iterables` so that we can decide if we want to be aware of memory performance issues or not.

> But in ProtoSchool tutorials, we don't really need to worry about this since our code challenges will use small files, so we can concatenate everything without worrying about performance.

Reviewer 2: ```suggestion
In ProtoSchool tutorials, we don't really need to worry about this since our code challenges use small files, so we can concatenate everything without worrying about performance.
```

A `Buffer` is just a raw collection of bytes and as such doesn't make any assumptions about encodings or the type of data it contains. However, if we know the file being retrieved is a plain text file such as a `.txt`, we can convert its buffered contents to a UTF-8 string (an interpretation of those raw bytes) by calling the JavaScript method `.toString()`.

> But since we have multiple buffers of the same file, we need to reassemble them (concatenate) into a single buffer before converting it into a string. The [`it-concat`](https://www.npmjs.com/package/it-concat) package can do just that: iterate over all of the chunks of the file and put them back together for us.

Reviewer: ```suggestion
But since we have multiple buffers making up a single file, we need to reassemble them (concatenate) into a single buffer before converting it into a string. The [`it-concat`](https://www.npmjs.com/package/it-concat) package can do just that: iterate over all of the chunks of the file and put them back together for us.
```

Reviewer 3: FWIW `it-concat` turns the buffers into a [`BufferList`](https://www.npmjs.com/package/bl), which you have to call `.slice` on to get an actual `Buffer`.

It's a good approach if you are going to doing lots of operations on the contents afterwards or appending new data, etc but if you want to avoid the conceptual overhead of Buffers vs BufferLists, you can just use [`it-to-buffer`](https://www.npmjs.com/package/it-to-buffer) instead.

Author: hey @achingbrain! Thank you for the input.

That's a good point, we'll switch to `it-to-buffer` then 👍 
Also, I just noticed that `it-concat` depends on [`bl`](https://www.npmjs.com/package/bl), so maybe it is intended for nodejs and not browers? @alanshaw 
If that's the case, then it's one more reason to switch to `it-to-buffer`.
